File: gm.py
#!/usr/bin/python3

import json, sys
from math import floor

TURNS_THRESHOLD = 10

EXCLUDED = ["smeargle", "pikachu"]

GMFILE = "gamemaster.json"
CPMFILE = "cpm.txt"

# The reference matchup we score every moveset against. Attacker and defender are
# both level 40 with IV 15, so the CPM cancels out of the attack/defense ratio and
# the ranking doesn't depend on the level we picked; the constants are here so the
# assumption is visible and easy to change.
ATTACKER_LEVEL = 40
DEFENDER_LEVEL = 40
DEFENDER_BASE_DEF = 100
IV = 15

# The trainer-battle damage bonus, part of the damage formula.
PVP_BONUS = 1.3

# A Protect Shield reduces a charged move to 1 damage, it doesn't nullify it.
SHIELDED_DAMAGE = 1

TYPE_ABBR = {
        'normal': 'nrm',
        'fighting': 'fig',
        'flying': 'fly',
        'poison': 'poi',
        'ground': 'grd',
        'rock': 'rck',
        'bug': 'bug',
        'ghost': 'gho',
        'steel': 'stl',
        'fire': 'fir',
        'water': 'wtr',
        'grass': 'grs',
        'electric': 'elc',
        'psychic': 'psy',
        'ice': 'ice',
        'dragon': 'drg',
        'dark': 'drk',
        'fairy': 'fai',
        }

with open(GMFILE) as f:
    gm = json.load(f)

# The CP multiplier for each level, one "<level> <cpm>" pair per line.
with open(CPMFILE) as f:
    cpm = { int(level): float(mult) for level, mult in (line.split() for line in f) }

defense = (DEFENDER_BASE_DEF + IV) * cpm[DEFENDER_LEVEL]

moves = { move["moveId"]: move for move in gm["moves"] if not move.get("unlisted") == True }

# Here we filter out any Pokemon that we want to exclude.
pokemon = {
        mon["speciesId"]: mon

        for mon in gm["pokemon"]

        if mon["released"]
            and ((not "tags" in mon)
                or ("shadow" not in mon["tags"] 
                    and "mega" not in mon["tags"]))
}

EXCLUDED_DEX_NUMBERS = [
        pokemon[excl_id]["dex"]
        for excl_id in EXCLUDED
        ]

# These print statements are exploratory and should be turned off "in production"
if True:
    print([ mon for mon in gm["pokemon"] if mon["speciesId"].startswith("maushold") ], file=sys.stderr)

    print(gm.keys(), file=sys.stderr)
    print(gm["moves"][0].keys(), file=sys.stderr)

    #print(moves.keys(), file=sys.stderr)
    print(moves["THUNDER_SHOCK"], file=sys.stderr)
    print(moves["RETURN"], file=sys.stderr)
    print(gm["pokemon"][0], file=sys.stderr)
    gn = [ mon for mon in gm["pokemon"] if mon["speciesId"].startswith("greninja") ]
    for x in gn: print(x, file=sys.stderr)

    #print([ mon for mon in gm["pokemon"] if "family" in mon and mon["family"]["id"] == "FAMILY_RATTATA" ], file=sys.stderr)

mon_fastest_movesets = []

def move_damage(move, move_stab, attack):
    # The game's damage formula. Type effectiveness is deliberately left out: it's
    # situational, and we show the move types so players can account for it. Note
    # that the flooring and the +1 happen per hit, so this is not a simple scaling
    # of the move's power -- low-power fast moves gain proportionally more.
    return floor(0.5 * PVP_BONUS * move["power"] * move_stab * attack / defense) + 1

def do_move_cycle(fm, fm_stab, cm, cm_stab, attack, residual_energy, block):
    turns = 0
    energy = residual_energy
    damage = 0
    # Apply fast moves until enough energy is generated to fire a charged move
    while energy < cm["energy"]:
        turns += fm["turns"]
        energy += fm["energyGain"]
        damage += move_damage(fm, fm_stab, attack)
    # Apply the charged move. A blocked move still costs its energy and its turn.
    if block:
        damage += SHIELDED_DAMAGE
    else:
        damage += move_damage(cm, cm_stab, attack)
    turns += 1
    energy -= cm["energy"]
    return (turns, damage, energy)

def stab(mon, move):
    if move["type"] in mon["types"]:
        return 1.2
    else:
        return 1.0

def do_move_cycles(mon, fm, cm, cycle_count, block):
    fm_stab = stab(mon, fm)
    cm_stab = stab(mon, cm)
    attack = (mon["baseStats"]["atk"] + IV) * cpm[ATTACKER_LEVEL]
    turns = 0
    damage = 0
    residual_energy = 0
    blocks_left = 2 if block else 0
    for i in range(cycle_count):
        block_next = False
        if blocks_left > 0:
            block_next = True
            blocks_left -= 1
        (t, d, e) = do_move_cycle(fm, fm_stab, cm, cm_stab, attack,
                                  residual_energy, block_next)
        turns += t
        damage += d
        residual_energy = e
    return (turns, damage)

def calc_damage(mon, fm, cm, block):
    # We calculate damage across three charged moves. This allows us to account
    # for different move counts due to residual energy left over from the first
    # move.
    (turns, damage) = do_move_cycles(mon, fm, cm, 3, block)
    # The result is damage per turn, in HP, against our reference defender.
    return damage / turns

for mon in pokemon.values():
    if mon["dex"] in EXCLUDED_DEX_NUMBERS:
        continue
    fast_moves = [ moves[m] for m in mon["fastMoves"] if m in moves  and moves[m]["energyGain"]>0 ]
    chrg_moves = [ moves[m] for m in mon["chargedMoves"] if m in moves ]
    eliteMoves = mon.get("eliteMoves") or []
    movesets = []
    for fm in fast_moves:
        for cm in chrg_moves:
            turns = (cm["energy"] / fm["energyGain"]) * fm["turns"]
            if turns <= TURNS_THRESHOLD:
                damage = calc_damage(mon, fm, cm, False)
                damage_with_blocks = calc_damage(mon, fm, cm, True)
                movesets.append({
                    "mon": mon,
                    "turns": turns,
                    "fm": fm,
                    "cm": cm,
                    "fm_elite": fm["moveId"] in eliteMoves,
                    "cm_elite": cm["moveId"] in eliteMoves,
                    "damage": damage,
                    "damage_with_blocks": damage_with_blocks,
                    })
    if len(movesets) == 0:
        continue
    movesets.sort(key = lambda ms: ms["turns"])
    best_turns = movesets[0]["turns"]
    for ms in movesets:
        ms["best"] = (ms["turns"] == best_turns)
    mon_fastest_movesets.extend(movesets)

def sort_key(moveset):
    return (moveset["turns"], -moveset["damage"], moveset["mon"]["speciesId"])
    #return "%05.2f--%d--%s" % (moveset["turns"], 1000000-moveset["damage"], moveset["mon"]["speciesId"])
mon_fastest_movesets.sort(key=sort_key)

count = 0
prev_turns = 0
for ms in mon_fastest_movesets:
    count += 1
    turns = ms["turns"]
    if turns > 10:
        break
    turns = round(turns, 2)
    if turns != prev_turns:
        print()
        print("TURNS:", turns)
        print()
        prev_turns = turns
    (fmt, cmt) = ( TYPE_ABBR[ms[x]["type"]] for x in ["fm","cm"] )
    print("%s" % ms["mon"]["speciesName"], end='')
    if not ms["best"]:
        print("*", end='')
    print(": %s%s [%s] / %s%s [%s];" % (
              '*' if ms["fm_elite"] else '',
              ms["fm"]["name"],
              fmt,
              '*' if ms["cm_elite"] else '',
              ms["cm"]["name"],
              cmt),
          "Turns: %s;" % str(turns),
          "Damage: %.2f;" % ms["damage"],
          "Damage (blocks): %.2f;" % ms["damage_with_blocks"],
          end=''
          )
    if fmt == cmt:
        print(" (both %s)" % ms["fm"]["type"], end='')
    print()

File: gm.js
"use strict";

// Port of gm.py: finds the Pokemon with the fastest charged-move cycles for
// Pokemon Go PVP, loading the gamemaster data live from pvpoke.

const GM_URL =
  "https://raw.githubusercontent.com/pvpoke/pvpoke/refs/heads/master/src/data/gamemaster.json";

const DEFAULT_TURNS_THRESHOLD = 10;

const TYPE_ABBR = {
  normal: "nrm",
  fighting: "fig",
  flying: "fly",
  poison: "poi",
  ground: "grd",
  rock: "rck",
  bug: "bug",
  ghost: "gho",
  steel: "stl",
  fire: "fir",
  water: "wtr",
  grass: "grs",
  electric: "elc",
  psychic: "psy",
  ice: "ice",
  dragon: "drg",
  dark: "drk",
  fairy: "fai",
};

const TYPE_COLORS = {
  normal: "#A8A77A",
  fighting: "#C22E28",
  flying: "#A98FF3",
  poison: "#A33EA1",
  ground: "#E2BF65",
  rock: "#B6A136",
  bug: "#A6B91A",
  ghost: "#735797",
  steel: "#B7B7CE",
  fire: "#EE8130",
  water: "#6390F0",
  grass: "#7AC74C",
  electric: "#F7D02C",
  psychic: "#F95587",
  ice: "#96D9D6",
  dragon: "#6F35FC",
  dark: "#705746",
  fairy: "#D685AD",
};

// ---------------------------------------------------------------- computation

// Matches Python's "%05.2f" / str(round(x, 2)): always at least one decimal,
// so 8 renders as "8.0" and 7.5 as "7.5".
function formatTurns(turns) {
  return turns.toFixed(2).replace(/0$/, "");
}

function indexGamemaster(gm) {
  const moves = new Map();
  for (const move of gm.moves) {
    if (move.unlisted !== true) moves.set(move.moveId, move);
  }

  const pokemon = [];
  for (const mon of gm.pokemon) {
    const tags = mon.tags;
    if (tags && (tags.includes("shadow") || tags.includes("mega"))) continue;
    if (mon.speciesId === "smeargle") continue; // learns everything; not useful
    pokemon.push(mon);
  }

  return { moves, pokemon };
}

function findMovesets({ moves, pokemon }, turnsThreshold) {
  const results = [];

  for (const mon of pokemon) {
    const fms = mon.fastMoves
      .map((m) => moves.get(m))
      .filter((m) => m && m.energyGain > 0);
    const cms = mon.chargedMoves.map((m) => moves.get(m)).filter((m) => m);

    const movesets = [];
    for (const fm of fms) {
      for (const cm of cms) {
        const turns = (cm.energy / fm.energyGain) * fm.turns;
        if (turns > turnsThreshold) continue;

        let fmPower = (fm.power * turns) / (turns + 1);
        let cmPower = cm.power / (turns + 1);
        if (mon.types.includes(fm.type)) fmPower *= 1.25; // STAB
        if (mon.types.includes(cm.type)) cmPower *= 1.25; // STAB

        movesets.push({
          mon,
          fm,
          cm,
          turns,
          damage: (fmPower + cmPower) * mon.baseStats.atk,
        });
      }
    }
    if (movesets.length === 0) continue;

    // Flag the moveset(s) that give this Pokemon its shortest cycle.
    const bestTurns = Math.min(...movesets.map((ms) => ms.turns));
    for (const ms of movesets) ms.best = ms.turns === bestTurns;

    results.push(...movesets);
  }

  // Fastest first, then hardest-hitting, then by species for a stable order.
  for (const ms of results) {
    ms.sortKey =
      ms.turns.toFixed(2).padStart(5, "0") +
      "--" +
      Math.trunc(1000000 - ms.damage) +
      "--" +
      ms.mon.speciesId;
  }
  results.sort((a, b) => (a.sortKey < b.sortKey ? -1 : a.sortKey > b.sortKey ? 1 : 0));

  return results;
}

// ------------------------------------------------------------------ rendering

function typeBadge(type) {
  const span = document.createElement("span");
  span.className = "type";
  span.textContent = TYPE_ABBR[type] || type;
  span.title = type;
  const color = TYPE_COLORS[type] || "#888";
  span.style.background = color;
  span.style.color = isLight(color) ? "#1a1a1a" : "#fff";
  return span;
}

function isLight(hex) {
  const n = parseInt(hex.slice(1), 16);
  const [r, g, b] = [(n >> 16) & 255, (n >> 8) & 255, n & 255];
  // Rec. 709 relative luminance.
  return 0.2126 * r + 0.7152 * g + 0.0722 * b > 150;
}

function moveCell(move, mon) {
  const td = document.createElement("td");
  td.className = "move";
  const name = document.createElement("span");
  name.textContent = move.name;
  if (mon.types.includes(move.type)) {
    name.className = "stab";
    name.title = "STAB";
  }
  td.append(name, " ", typeBadge(move.type));
  return td;
}

function render(movesets, opts) {
  const tbody = document.getElementById("results");
  tbody.replaceChildren();

  const search = opts.search.trim().toLowerCase();
  const shown = movesets.filter((ms) => {
    if (opts.bestOnly && !ms.best) return false;
    if (!search) return true;
    return (
      ms.mon.speciesName.toLowerCase().includes(search) ||
      ms.fm.name.toLowerCase().includes(search) ||
      ms.cm.name.toLowerCase().includes(search) ||
      ms.fm.type.includes(search) ||
      ms.cm.type.includes(search)
    );
  });

  let prevTurns = null;
  const frag = document.createDocumentFragment();

  for (const ms of shown) {
    const label = formatTurns(ms.turns);
    if (label !== prevTurns) {
      const tr = document.createElement("tr");
      tr.className = "group";
      const td = document.createElement("td");
      td.colSpan = 5;
      td.textContent = "TURNS: " + label;
      tr.append(td);
      frag.append(tr);
      prevTurns = label;
    }

    const tr = document.createElement("tr");
    if (!ms.best) tr.className = "notbest";

    const name = document.createElement("td");
    name.className = "mon";
    name.textContent = ms.mon.speciesName;
    if (!ms.best) {
      const star = document.createElement("span");
      star.className = "star";
      star.textContent = "*";
      star.title = "Not this Pokemon's fastest moveset";
      name.append(star);
    }

    const turns = document.createElement("td");
    turns.className = "num";
    turns.textContent = label;

    const damage = document.createElement("td");
    damage.className = "num";
    damage.textContent = Math.trunc(ms.damage).toLocaleString();

    tr.append(name, moveCell(ms.fm, ms.mon), moveCell(ms.cm, ms.mon), turns, damage);
    frag.append(tr);
  }

  tbody.append(frag);
  document.getElementById("count").textContent =
    shown.length.toLocaleString() + " moveset" + (shown.length === 1 ? "" : "s");
}

// --------------------------------------------------------------------- wiring

function setStatus(text, isError) {
  const el = document.getElementById("status");
  el.textContent = text;
  el.classList.toggle("error", !!isError);
}

async function main() {
  const thresholdEl = document.getElementById("threshold");
  const bestOnlyEl = document.getElementById("bestOnly");
  const searchEl = document.getElementById("search");

  setStatus("Loading gamemaster.json from pvpoke…");

  let gm;
  try {
    const resp = await fetch(GM_URL, { cache: "no-cache" });
    if (!resp.ok) throw new Error("HTTP " + resp.status + " " + resp.statusText);
    gm = await resp.json();
  } catch (err) {
    setStatus(
      "Could not load the gamemaster: " +
        err.message +
        ". If you opened this file directly, try serving it instead " +
        "(python3 -m http.server) so the browser allows the request.",
      true
    );
    return;
  }

  const indexed = indexGamemaster(gm);
  const updated = gm.timestamp ? new Date(gm.timestamp).toLocaleString() : "unknown";
  setStatus(
    `${indexed.pokemon.length.toLocaleString()} Pokemon and ` +
      `${indexed.moves.size.toLocaleString()} moves · gamemaster updated ${updated}`
  );

  let movesets = [];
  const recompute = () => {
    const threshold = Number(thresholdEl.value) || DEFAULT_TURNS_THRESHOLD;
    movesets = findMovesets(indexed, threshold);
    redraw();
  };
  const redraw = () =>
    render(movesets, { bestOnly: bestOnlyEl.checked, search: searchEl.value });

  thresholdEl.addEventListener("change", recompute);
  bestOnlyEl.addEventListener("change", redraw);
  searchEl.addEventListener("input", redraw);

  recompute();
}

document.addEventListener("DOMContentLoaded", main);
